content/doc/book/pipeline/jenkinsfile.adoc

```diff
@@ -341,6 +341,16 @@ I said, Hello Mr. Jenkins
 Understanding how to use string interpolation is vital for using some of
 Pipeline's more advanced features.
 
+==== [red]*WARNING: Interpolation of Sensitive Variables*
+Groovy string interpolation should [red]*never* be used with sensitive variables. This is
+because the sensitive variable will be interpolated at Groovy evaluation and the variable's
+value could possibly be made available earlier than intended, resulting in the sensitive data leaking.
+
+For example, consider a password variable passed to the shell step. Should Groovy perform the
+interpolation, the password value would become available not only in the shell step, but the
+Jenkins Controller as well. Additionally, if the password contained special characters
+(e.g. ' " / \ $ & % ^ > < | ;) it would also be possible to perform SQL-like injections into the
+shell step. Special characters can also prevent the password from being masked properly in the console logs.
```

Comment:
I would probably split this into a separate section, since it applies to any user-controlled parameter or environment variable, not just secrets, and it is specific to the `sh`, `bat`, `powershell`, and `pwsh` steps, and could also use an example or two to demonstrate the problem and the possible solutions.

Reply:
split in 5e8a114acd559b02df299446e548619b622f724a